nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DeleteFile.java

```diff
@@ -0,0 +1,138 @@
+
+    public static final PropertyDescriptor DIRECTORY_PATH = new PropertyDescriptor.Builder()
+            .name("Directory Path")
+            .displayName("Directory Path")
+            .description("The path to the directory the file to delete is located in.")
+            .required(true)
+            .defaultValue("${" + CoreAttributes.ABSOLUTE_PATH.key() + "}")
+            .addValidator(StandardValidators.createDirectoryExistsValidator(true, false))
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+            .build();
+    public static final PropertyDescriptor FILENAME = new PropertyDescriptor.Builder()
+            .name("Filename")
+            .displayName("Filename")
+            .description("The name of the file to delete.")
+            .required(true)
+            .defaultValue("${" + CoreAttributes.FILENAME.key() + "}")
+            .addValidator(StandardValidators.NON_EMPTY_EL_VALIDATOR)
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+            .build();
+
+    private final static List<PropertyDescriptor> properties = List.of(DIRECTORY_PATH, FILENAME);
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final String directoryPathProperty = context.getProperty(DIRECTORY_PATH).evaluateAttributeExpressions(flowFile).getValue();
+        final String filename = context.getProperty(FILENAME).evaluateAttributeExpressions(flowFile).getValue();
+
+        try {
+            final Path directoryPath = Paths.get(directoryPathProperty).toRealPath();
+            final Path filePath = directoryPath.resolve(filename).toRealPath();
+
+            if (!directoryPath.equals(filePath.getParent())) {
+                getLogger().error("Attempting to delete file at path '{}' which is not a direct child of the directory '{}'", filePath, directoryPath);
+                session.penalize(flowFile);
+                session.transfer(flowFile, REL_FAILURE);
+                return;
+            }
+
+            Files.delete(filePath);
+
+            session.transfer(flowFile, REL_SUCCESS);
+        } catch (NoSuchFileException noSuchFileException) {
+            session.transfer(flowFile, REL_NOT_FOUND);
+        } catch (IOException ioException) {
+            getLogger().error("Failed to delete file '{}' in directory '{}'", filename, directoryPathProperty, ioException);
+
+            session.penalize(flowFile);
+            session.transfer(flowFile, REL_FAILURE);
```

Comment:
For each place that transfers to REL_FAILURE, I think it would be helpful to add to the flowFile a writesAttribute with an error message about why it failed to help humans troubleshooting a failure. For this case of REL_FAILURE I would use wording like the log error on line 132 and add the exception class name and the exception,getMessage(). I suggest writing an error attribute for line 122 above also.

Reply:
@jrsteinebrey  We should not start having a standard of writing exception class names in flowfile attributes nor should we make it a standard to have raw exception messages in an attribute.  

I certainly agree we can and should do more to provide more context to the user but doing it this way is not nearly as effective as the component/extension developer being thoughtful about which exceptions can occur and what would be best to convey to the user both in terms of what we log, what we may add as context in flowfile attributes, and what relatioships we give them to make different flow design decisions.

Reply:
@joewitt @EndzeitBegins 
OK, Joe. I retract my suggestion based on your points.